src/backend/distributed/commands/multi_copy.c

```diff
@@ -763,7 +763,7 @@ OpenCopyConnections(CopyStmt *copyStatement, ShardConnections *shardConnections,
 		ShardPlacement *placement = (ShardPlacement *) lfirst(placementCell);
 		char *nodeUser = CurrentUserName();
 		MultiConnection *connection = NULL;
-		uint32 connectionFlags = FOR_DML;
+		uint32 connectionFlags = FOR_DML | CONNECTION_PER_PLACEMENT;
```

Comment:
copy ... to ... goes through the real-time executor, which does not use the connection API, so it would not use the right connections and thus would not see preceding writes, or would form a self-deadlock in case of DDL.